keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

```diff
@@ -0,0 +1,1114 @@
+type Device struct {
+    // Name is unique identifier among all devices managed by
+    // the driver in the pool. It must be a DNS label.
+    //
+    // +required
+    Name string
+
+    // Basic defines one device instance.
+    //
+    // +optional
+    // +oneOf=deviceType
+    Basic *BasicDevice
+}
+
+// BasicDevice represents a basic device instance.
+type BasicDevice struct {
+    // Attributes defines the set of attributes for this device.
+    // The name of each attribute must be unique in that set.
+    //
+    // The maximum number of attributes and capacities combined is 32.
+    //
+    // +optional
+    Attributes map[QualifiedName]DeviceAttribute
+
+    // BindingGates defines the gates for binding.
+    //
+    // +optional
+    BindingGates []string
+
+    // BindingFailureGates defines the gates for binding failure.
+    //
+    // +optional
+    BindingFailureGates []string
+
+    // UsageRestrictedToNode indicates if the usage of an allocation involving this device
+    // has to be limited to exactly the node that was chosen when allocating the claim.
+    //
+    // +optional
+    UsageRestrictedToNode bool
+}
+```
+
+#### AllocatedDeviceStatus Enhancements
+
+The `BindingGates` and `BindingFailureGates` fields within `AllocatedDeviceStatus` are used to indicate the status of the device attachment.
+These fields will contain a list of conditions, each representing a specific state or event related to the device.
+
+For this feature, following fields are added:
+
+```go
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+    ...
+    // BindingGates defines the gates for binding.
+    //
+    // +optional
+    BindingGates map[string]bool
+
+    // BindingFailureGates defines the gates for binding failure.
+    //
+    // +optional
+    BindingFailureGates map[string]bool
+}
+```
+
+#### Scheduler DRA Plugin Modifications
+
+When `UsageRestrictedToNode: true` is set, the scheduler DRA plugin will perform the following steps:
+
+1. **Set NodeSelector**: Before the `PreBind` phase, add the `NodeName` to the `ResourceClaim`'s `NodeSelector`.
+
+If Gates are present, the scheduler DRA plugin will perform the following steps during the `PreBind` phase:
+
+2. **Copy Gates**: Copy `BindingGates` and `BindingFailureGates` from `ResourceSlice.Device.Basic` to `AllocatedDeviceStatus`.
+3. **Wait for Conditions**: Wait for the following conditions:
+   - If `NeedToPreparing` is `True`, wait until `IsPrepared` is `True` before proceeding to Bind.
+   - If `PreparingFailed` is `True`, clear the allocation in the `ResourceClaim` and reschedule the Pod.
+   - If the preparation takes longer than the `PreparingTimeout` period, clear the allocation in the `ResourceClaim` and reschedule the Pod.
+
+To support these steps, the following keys are defined:
+
+```go
+const (
```

Comment:
These are all now driver-defined and not K8s things. There should be no need for them here.

If you want to include an *example* of how driver's could use the new fields to manage an attachment process, that would be great. But it's an example only.